src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup_new.tsx

```diff
@@ -0,0 +1,254 @@
+          targetSubjectIds,
+          fromDate,
+          toDate,
+        };
+        const firstPageResponse = await getExportDataApi(body, abortControllerRef.current.signal);
+        const { result: firstPageData, count = 0 } = firstPageResponse.data;
+        const pageLimit = getExportPageAmount(count);
+        setExportDataQueue([firstPageData]);
+
+        if (pageLimit > 1) {
+          limitRef.current = pageLimit;
+          for (let page = 2; page <= pageLimit; page++) {
+            const nextPageResponse = await getExportDataApi(
+              {
+                ...body,
+                page,
+              },
+              abortControllerRef.current.signal,
+            );
+            const { result: nextPageData } = nextPageResponse.data;
+
+            setExportDataQueue((prevState) =>
+              prevState ? [...prevState, nextPageData] : [nextPageData],
+            );
+          }
+        }
+      } catch (e) {
+        const error = e as TypeError;
+        console.warn('Error while downloading export data', error);
+        setActiveModal(Modals.ExportError);
+        await sendLogFile({ error });
+      } finally {
+        dataExportingApiRef.current = false;
+      }
+    },
+    [getValues, dataExportingApiRef, limitRef, setDataIsExporting, setExportDataQueue],
+  );
+  const handleRetry = () => {
+    setActiveModal(Modals.DataExport);
+    handleDataExportSubmit();
+  };
+
+  const percentages = Math.floor(
+    (Math.max(...[finishedPagesRef.current.size, 1]) / limitRef.current) * 100,
+  );
+  const percentagesBuffer = Math.min(100, percentages + BUFFER_PERCENTAGE);
+  const showExportPercentages = limitRef.current > 1;
+
+  const renderDataExportContent = () => {
+    if (dataIsExporting) {
+      return (
+        <>
+          <StyledBodyLarge sx={{ margin: theme.spacing(-2.4, 0, 2.4) }}>
+            {t('waitForRespondentDataDownload')}
+            {showExportPercentages && (
+              <>
+                <br />
+                <br />
+                {t('dataProcessing', {
+                  percentages,
+                })}
+              </>
+            )}
+          </StyledBodyLarge>
+          <StyledLinearProgress
+            variant={showExportPercentages ? 'buffer' : 'indeterminate'}
+            value={percentages}
+            valueBuffer={percentagesBuffer}
+          />
+        </>
+      );
+    }
+
+    return (
+      <>
+        <StyledBodyLarge sx={{ margin: theme.spacing(-2.4, 0, 2.4) }}>
+          {t('selectAppletToExportRespondentsData')}
+        </StyledBodyLarge>
+        <AppletsSmallTable tableRows={tableRows} />
+      </>
+    );
+  };
+
+  useEffect(() => {
+    setActiveModal(showEnterPwdScreen ? Modals.PasswordCheck : Modals.DataExport);
+  }, [showEnterPwdScreen]);
+
+  useEffect(
+    () => () => {
+      if (abortControllerRef.current) {
+        // stop requests to the API if the component unmounts
+        abortControllerRef.current.abort();
```

Comment:
This is a good example of using `useEffect` as it syncs with the external API (AbortController). However, the right approach would be to hide the aborController in a class instance AND invoke a special method that describes itself. For example:
```
useEffect(
    () => () => {
      ExportDataService.cancel()
    }
  [],
);
```